Self-Consistent-Field/RHF_symmetry.py

```diff
@@ -0,0 +1,186 @@
+t = time.time()
+mints = psi4.core.MintsHelper(wfn.basisset())
+
+nirrep = wfn.nirrep()
+# Water belongs to the C2v point group, which has four irreducable
+# representations (irreps): A_1, A_2, B_1, and B_2.
+assert nirrep == 4
+
+dimension_to_list = lambda dim: [dim[i] for i in range(nirrep)]
+nsopi = dimension_to_list(wfn.nsopi())
+
+# Get nbf and ndocc for closed shell molecules
+nbf = sum(nsopi)
+ndocc = wfn.nalpha()
+
+print('\nNumber of occupied orbitals: %d' % ndocc)
+print('Number of basis functions: %d' % nbf)
+print('Number of spin orbitals per irrep:', nsopi)
+
+# Run a quick check to make sure everything will fit into memory
+I_Size = (nbf**4) * 8.e-9
+print("\nSize of the ERI tensor will be %4.2f GB." % I_Size)
+
+# Estimate memory usage
+memory_footprint = I_Size * 1.5
+if I_Size > numpy_memory:
+    psi4.core.clean()
+    raise Exception("Estimated memory utilization (%4.2f GB) exceeds numpy_memory "
+                    "limit of %4.2f GB." % (memory_footprint, numpy_memory))
+
+# A matrix is returned for every irrep, even if the irrep is not present for
+# the molecule; these must be filtered out to avoid problems with NumPy arrays
+# that have a zero dimension.
+filter_empty_irrep = lambda coll: tuple(m for m in coll if all(m.shape))
+
+# The convention will be to have quantities in the spin-orbital basis ending
+# with an underscore, each consisting of one matrix per irrep.
+S_ = filter_empty_irrep(mints.so_overlap().to_array())
+T_ = filter_empty_irrep(mints.so_kinetic().to_array())
+V_ = filter_empty_irrep(mints.so_potential().to_array())
+# The two-electron integrals are not blocked according to symmetry, so a
+# transformation between the AO and SO bases will be required.
+I_AO = np.asarray(mints.ao_eri())
+transformers_ = filter_empty_irrep(wfn.aotoso().to_array())
+
+assert len(S_) == len(T_) == len(V_) == len(transformers_)
+nirrep = len(S_)
+nsopi = [n for n in nsopi if n > 0]
+
+print('\nTotal time taken for integrals: %.3f seconds.' % (time.time() - t))
+t = time.time()
+
+# Build H_core: [Szabo:1996] Eqn. 3.153, pp. 141
+H_ = [T + V for (T, V) in zip(T_, V_)]
+
+# Build the orthogonalizer A = S^(-1/2) for each irrep.
+A_ = []
+for S in S_:
+    # Application of a function to a matrix requires transforming to
+    # the diagonal basis, applying the function to the diagonal form,
+    # then backtransformation to the original basis.
+    eigval, eigvec = np.linalg.eigh(S)
+    eigval_diag = np.diag(eigval ** (-1./2))
+    A = eigvec.dot(eigval_diag).dot(eigvec.T)
+    A_.append(A)
+
+# Calculate initial core guess: [Szabo:1996] pp. 145
+C_ = []
+e_ = []
+for i in range(nirrep):
+    Hp = A_[i].dot(H_[i]).dot(A_[i])  # Eqn. 3.177
+    e, C2 = np.linalg.eigh(Hp)        # Solving Eqn. 1.178
+    C = A_[i].dot(C2)                 # Back transform, Eqn. 3.174
+    C_.append(C)
+    e_.append(e)
+
+# Initial occupations are taken from the lowest eigenvalues (energies)
+# of the guess coefficients.
```

Comment:
Toying around to make this simpler, perhaps we could:
 - concatenate the blocks
 - sort
 - find the value of the energy of the HOMO
 - build ndoccpi by `[np.sum(e_[I] < homo) for I in range(nirrep)]`

Or maybe build `[(e, i) for e in e_[I] for I in range(nirrep)]`, sort, and count?

Reply:
I give up on not making my browser convert `i` into caps.